src/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs

```diff
@@ -842,12 +881,26 @@ internal sealed class AwaitableSocketAsyncEventArgs : SocketAsyncEventArgs, IVal
 
             private void Release()
             {
+                _cancellationToken = default;
                 _token++;
                 Volatile.Write(ref _continuation, s_availableSentinel);
             }
 
             protected override void OnCompleted(SocketAsyncEventArgs _)
             {
+                // _cancellationState will be None if cancellation can't happen,
+                // Registering if it's in the process of being registered, and
+                // Registered if it's been registered.  If cancellation may happen,
+                // we spin until it's been registered, so that we can safely unregister.
+                // That way, there's no concern about leaking a registration.
+                if (_cancellationState != CancellationState_None)
+                {
+                    var sw = new SpinWait();
```

Comment:
would it make sense to call this only when needed? (unless it is really cheap). I would expect that cases when we have contention on state would be rare.

Reply:
It's cheap. SpinWait is a struct, so
```
var sw = new SpinWait();
```
is identical to:
```
SpinWait sw = default;
```